### PR TITLE
Implementation of Variable Length Packets

### PR DESCRIPTION
**Pull request by ddasilva**
This pull request implements variable length packets, via a custom decoder and interface through the `VariableLength` class. For convenience, I put a build of the API docs at this URL so reviewers can read:
https://danieldasilva.org/var_length/ccsdspy.html

This merge request makes the following assumptions, which I think should meet the majority of use cases that occur in practice. I'm curious about what @ehsteve thinks about these assumptions based the data they're using. It assumes:
1. There is one variable length field per packet,  ~~and it always comes at the end~~
2. It starts and ends cleanly on byte boundaries (it doesn't span an "odd" number of bits between bytes)

When you decode a variable length field, you get back an array of `dtype=object`, where each element is a reference to another array of the specified type (with different lengths each time). One thing to note is that this variable length decoder is slightly (but not impractically) slower than the `FixedLength` decoder, so we want to continue to support `FixedLength`. 

Since the front page is kind of long, and the rules for this class are worth spelling out, I'm going to add a new page on the sidebar for this (and then link to it inline with the Fixed Length discussion). But for now I want to focus on making sure the decoder is right.

### CODE REVIEW

**Comment by ddasilva** [comment]
Closes #24

**Comment by ehsteve** [comment]
I worry about your first assumption. It feels like common practice to put a checksum at the end of a packet or some other "footer" data.

**Comment by ddasilva** [comment]
That’s fine, I just didn’t know. I’ll update the PR

**Comment by ddasilva** [comment]
> I worry about your first assumption. It feels like common practice to put a checksum at the end of a packet or some other "footer" data.

@ehsteve I updated the pull request with this constraint removed, and the docs here have been updated
https://danieldasilva.org/var_length/ccsdspy.html

**Comment by ehsteve** [comment]
Wow! Amazing! I'll have a closer look in the new year. Happy new year!

**Comment by jmbhughes** [comment]
This should work great with PUNCH image packets I think. We have just one expanding data field which is (at the moment) always at the end.

**Comment by ddasilva** [comment]
Thanks @jmbhughes !

File: ccsdspy/tests/test_var_length.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    assert field_arrays["data"].dtype == object
+    assert field_arrays["data"].size == 10
+    assert np.all(field_arrays["footer"] == 1)
+
+    sizes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
+
+    for i in range(10):
+        assert field_arrays["data"][i].shape == (sizes[i],)
+        expected = np.arange(sizes[i], 2 * sizes[i])
```
Comment (ehsteve) [inline ccsdspy/tests/test_var_length.py:69]:
```suggestion
        expected = np.arange(sizes[i], dtype='uint16') + sizes[i]
```

File: ccsdspy/tests/test_var_length.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                bit_length=16,
+                array_shape="expand",
+            ),
+            PacketField(name="footer", data_type="uint", bit_length=16),
+        ]
+    )
+
+    field_arrays = pkt.load(bin_path, include_primary_header=include_primary_header)
+
+    assert field_arrays["data"].dtype == object
```
Comment (ehsteve) [inline ccsdspy/tests/test_var_length.py:61]: would be good to check if the dtype is indeed 16 bit long.

**Review by ehsteve — APPROVED** [review]
This looks great!

**Pull request merged**